livre/core/data/Histogram.h

```diff
@@ -0,0 +1,99 @@
+/* Copyright (c) 2011-2016, EPFL/Blue Brain Project
+ *                          Ahmet Bilgili <ahmet.bilgili@epfl.ch>
+ *
+ * This file is part of Livre <https://github.com/BlueBrain/Livre>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _Histogram_h_
+#define _Histogram_h_
+
+#include <livre/core/api.h>
+#include <livre/core/mathTypes.h>
+#include <co/distributable.h> // base class
+
+#include <lexis/render/viewHistogram.h> // base class
+
+
+namespace livre
+{
+
+namespace
+{
+const size_t binCount = 256;
+}
+
+class Histogram : public co::Distributable< ::lexis::render::ViewHistogram >
+{
+public:
+    /**
+     * Constructor. Initializes histogram with 0 values
+     */
+    LIVRECORE_API Histogram();
+
+    /**
+     * Copy the histogram.
+     * @param hist The histogram to be copied.
+     */
+    LIVRECORE_API Histogram( const Histogram& hist );
+    LIVRECORE_API Histogram& operator=( const Histogram& rhs );
+
+    /**
+     * @return true if total area of the view port related to histogram is ~1.0
+     */
+    LIVRECORE_API bool isComplete() const;
+
+    /**
+     * Computes the addition of two histograms and modifies the object
+     * @param histogram is the histogram to add
+     * @return the modified histogram
+     * @throw std::logic_error if ids of histograms does not match
+     */
+    LIVRECORE_API Histogram& operator+=( const Histogram& histogram );
+
+    /**
+     * Queries uniformity of the data. ( If histogram has only one value )
+     * @param uniformIndex is set with unfiorm index if data is uniform
+     * @return true data is composed of uniform values
+     */
+    LIVRECORE_API bool isUniform( float& uniformIndex ) const;
+
+    /**
+     * @return the index with the minimum value
+     */
+    LIVRECORE_API size_t getMinIndex() const;
+
+    /**
+     * @return the index with the maximum value
+     */
+    LIVRECORE_API size_t getMaxIndex() const;
+
+    /**
+     * @return the bin count.
+     */
+    static size_t getBinCount() { return binCount; }
+
+private:
+
+    mutable size_t _minIndex;
+    mutable size_t _maxIndex;
```

Comment:
Mutable?? This is fishy. If it because you compute then lazily?
With only 256 bins I think you can afford updating them every time the histogram is modified.